src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java

```diff
@@ -46,11 +102,229 @@ public void execute() throws MojoExecutionException, MojoFailureException {
+                    }
+                }
+                if (!found) {
+                    throw new MojoExecutionException("could not find dependency " + key);
+                }
+            }
+            Properties properties = project.getProperties();
+            getLog().info("Replacing POM-defined classpath elements " + classpathDependencyExcludes + " with " + additionalClasspathElements);
+            // cf. http://maven.apache.org/surefire/maven-surefire-plugin/test-mojo.html
+            properties.setProperty("maven.test.additionalClasspath", StringUtils.join(additionalClasspathElements, ','));
+            properties.setProperty("maven.test.dependency.excludes", StringUtils.join(classpathDependencyExcludes, ','));
+        }
     }
+
+    private Artifact replace(Artifact a, String version) throws MojoExecutionException {
+        Artifact a2 = new DefaultArtifact(a.getGroupId(), a.getArtifactId(), VersionRange.createFromVersion(version), a.getScope(), a.getType(), a.getClassifier(), a.getArtifactHandler(), a.isOptional());
+        try {
+            return artifactResolver.resolveArtifact(session.getProjectBuildingRequest(), a2).getArtifact();
+        } catch (ArtifactResolverException x) {
+            throw new MojoExecutionException("could not find " + a + " in version " + version + ": " + x, x);
+        }
+    }
+
+    // Adapted from RequireUpperBoundDeps @ 731ea7a693a0986f2054b6a73a86a31373df59ec. TODO delete extraneous stuff and simplify to the logic we actually need here:
+    private class RequireUpperBoundDepsVisitor implements DependencyNodeVisitor {
+
+        private boolean uniqueVersions;
+
+        public void setUniqueVersions(boolean uniqueVersions) {
+            this.uniqueVersions = uniqueVersions;
+        }
+
+        private Map<String, List<DependencyNodeHopCountPair>> keyToPairsMap
+                = new LinkedHashMap<String, List<DependencyNodeHopCountPair>>();
+
+        public boolean visit(DependencyNode node) {
+            DependencyNodeHopCountPair pair = new DependencyNodeHopCountPair(node);
+            String key = pair.constructKey();
+            List<DependencyNodeHopCountPair> pairs = keyToPairsMap.get(key);
+            if (pairs == null) {
+                pairs = new ArrayList<DependencyNodeHopCountPair>();
+                keyToPairsMap.put(key, pairs);
+            }
+            pairs.add(pair);
+            Collections.sort(pairs);
+            return true;
+        }
+
+        public boolean endVisit(DependencyNode node) {
+            return true;
+        }
+
+        // added for TestDependencyMojo in place of getConflicts/containsConflicts
+        @SuppressWarnings("unchecked")
+        public Map<String, String> upperBounds() {
+            Map<String, String> r = new HashMap<>();
+            // TODO this does not suffice; does not find that workflow-api needs to go from 2.11 to 2.16, presumably because it was not a direct dependency to begin with
```

Comment:
I think to resolve this we would need to mutate the shadow `Project` with the results of this method call, then re-do dependency resolution (i.e., do the equivalent of `LifecycleDependencyResolver#getDependencies`), keeping track of the versions before and after (re-)resolution, then apply the difference to `overrideVersions` map.

Reply:
> mutate the shadow Project with the results of this method call

To clarify, I mean doing this the same way you would if editing the POM manually: in the case of a direct dependency, by adjusting the version; in the case of a transitive dependency, by adding a new entry to the dependency management section.

Reply:
Hmm. Sounds tricky but could work. (Bringing this PR over the finish line would require some dedicated attention, and in general my workday only allows for a few minutes here and there on Jenkins infra stuff.) Thanks for helping think about it.